### PR TITLE
fix: upgrade process to version-14 when currency opportunity was not set

### PR DESCRIPTION
**Pull request by FHenry**
I encounter Opportunity without currency on customer erpnext instance, It make the upgrade crash

```
frappe@be524147e466:/workspace/development/frappe-bench$ bench --site erpnext14test.local migrate
Migrating erpnext14test.local
Executing execute:frappe.reload_doc('core', 'doctype', 'doctype_state', force=True) #2021-12-15 in erpnext14test.local (erpnext14test)
Success: Done in 0.137s
Executing execute:frappe.reload_doc('core', 'doctype', 'doctype') #2022-06-21 in erpnext14test.local (erpnext14test)
Success: Done in 0.425s
Executing frappe.patches.v13_0.jinja_hook in erpnext14test.local (erpnext14test)
Success: Done in 0.016s
Executing frappe.patches.v13_0.update_notification_channel_if_empty in erpnext14test.local (erpnext14test)
Success: Done in 0.267s
Executing execute:frappe.reload_doc('custom', 'doctype', 'custom_field') in erpnext14test.local (erpnext14test)
Success: Done in 0.244s
Executing frappe.patches.v14_0.update_workspace2 # 20.09.2021 in erpnext14test.local (erpnext14test)
Success: Done in 2.395s
Executing frappe.patches.v14_0.save_ratings_in_fraction #23-12-2021 in erpnext14test.local (erpnext14test)
Success: Done in 0.366s
Executing frappe.patches.v14_0.transform_todo_schema in erpnext14test.local (erpnext14test)
Success: Done in 0.219s
Executing frappe.patches.v14_0.remove_post_and_post_comment in erpnext14test.local (erpnext14test)
Success: Done in 0.145s
Executing frappe.patches.v14_0.reset_creation_datetime in erpnext14test.local (erpnext14test)
	Resetting creation datetimes for DocTypes
Success: Done in 0.154s
Executing frappe.patches.v14_0.remove_is_first_startup in erpnext14test.local (erpnext14test)
Success: Done in 0.016s
Executing frappe.patches.v14_0.clear_long_pending_stale_logs in erpnext14test.local (erpnext14test)
	Due to large size of log tables on old sites some table cleanups never finished during daily log clean up. This patch discards such data by using "big delete" code.

		ref: https://github.com/frappe/frappe/issues/16971

Clearing old Email Queue records
Clearing old Email Queue Recipient records
Success: Done in 0.666s
Executing frappe.patches.v14_0.log_settings_migration in erpnext14test.local (erpnext14test)
Success: Done in 0.343s
Executing frappe.patches.v14_0.setup_likes_from_feedback in erpnext14test.local (erpnext14test)
Success: Done in 0.22s
Executing frappe.patches.v14_0.update_webforms in erpnext14test.local (erpnext14test)
Success: Done in 0.401s
Executing frappe.patches.v14_0.delete_payment_gateways in erpnext14test.local (erpnext14test)
Success: Done in 0.356s
Executing erpnext.patches.v14_0.change_is_subcontracted_fieldtype in erpnext14test.local (erpnext14test)
Success: Done in 2.776s
Executing erpnext.patches.v13_0.add_doctype_to_sla #14-06-2021 in erpnext14test.local (erpnext14test)
Success: Done in 0.311s
Executing erpnext.patches.v13_0.update_response_by_variance in erpnext14test.local (erpnext14test)
Success: Done in 0.017s
Executing erpnext.patches.v14_0.update_opportunity_currency_fields in erpnext14test.local (erpnext14test)
Updating Dashboard for frappe
Updating Dashboard for erpnext
Updating Dashboard for cefpolymeres
Updating Dashboard for erpnext_france
Updating Dashboard for special_item_accountancy_code
Updating Dashboard for jobcard_planning
Updating customizations for Address
Updating customizations for Contact
Building search index for erpnext14test.local
Retrieving Routes                   : [========================================] 100%
Building Index                      : [========================================] 100%
Traceback (most recent call last):
  File "/home/frappe/.pyenv/versions/3.10.5/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/frappe/.pyenv/versions/3.10.5/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/workspace/development/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 109, in <module>
    main()
  File "/workspace/development/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 18, in main
    click.Group(commands=commands)(prog_name="bench")
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/workspace/development/frappe-bench/apps/frappe/frappe/commands/__init__.py", line 29, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/workspace/development/frappe-bench/apps/frappe/frappe/commands/site.py", line 523, in migrate
    SiteMigration(
  File "/workspace/development/frappe-bench/apps/frappe/frappe/migrate.py", line 172, in run
    self.run_schema_updates()
  File "/workspace/development/frappe-bench/apps/frappe/frappe/migrate.py", line 40, in wrapper
    ret = method(*args, **kwargs)
  File "/workspace/development/frappe-bench/apps/frappe/frappe/migrate.py", line 109, in run_schema_updates
    frappe.modules.patch_handler.run_all(
  File "/workspace/development/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 75, in run_all
    run_patch(patch)
  File "/workspace/development/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 62, in run_patch
    if not run_single(patchmodule=patch):
  File "/workspace/development/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 150, in run_single
    return execute_patch(patchmodule, method, methodargs)
  File "/workspace/development/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 185, in execute_patch
    _patch()
  File "/workspace/development/frappe-bench/apps/erpnext/erpnext/patches/v14_0/update_opportunity_currency_fields.py", line 27, in execute
    frappe.db.set_value(
  File "/workspace/development/frappe-bench/apps/frappe/frappe/database/database.py", line 896, in set_value
    query.run(debug=debug)
  File "/workspace/development/frappe-bench/apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "/workspace/development/frappe-bench/apps/frappe/frappe/database/database.py", line 209, in sql
    self._cursor.execute(query, values)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/workspace/development/frappe-bench/env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.IntegrityError: (1048, "Column 'conversion_rate' cannot be null")
```

Sample of a Opporuny record that make it crash

```
[
  {
    "name": "CRM-OPP-2021-00065",
    "creation": "2021-07-30 11:51:24.602005",
    "modified": "2021-11-18 12:04:07.605473",
    "modified_by": "user@test.com",
    "owner": "user@test.com",
    "docstatus": 0,
    "parent": null,
    "parentfield": null,
    "parenttype": null,
    "idx": 0,
    "naming_series": "CRM-OPP-.YYYY.-",
    "opportunity_from": "Lead",
    "party_name": "PROSPYYYY-0040",
    "customer_name": "MEGAPAK",
    "title": "MEGAPAK",
    "opportunity_type": "Ventes",
    "status": "Open",
    "converted_by": null,
    "sales_stage": "Analyse des besoins",
    "order_lost_reason": null,
    "mins_to_first_response": 2.000000,
    "contact_by": "user@test.com",
    "contact_date": "2022-09-21 10:56:48.0",
    "to_discuss": "MEGAPAK - M XXXX XXXX - 85 P\nREPONSE DE L'acheteur M XXXXX  : ..........  - ok mail",
    "currency": "",
    "opportunity_amount": 21221.000000000,
    "with_items": 0,
    "probability": 50.000000000,
    "customer_address": "MEGAPAK-Facturation",
    "address_display": null,
    "territory": "OUEST",
    "customer_group": null,
    "contact_person": null,
    "contact_display": "M XXXX XXXX",
    "contact_email": "contact@text.com",
    "contact_mobile": null,
    "source": "Prospection Tel",
    "campaign": null,
    "company": "COMPANY TEST",
    "transaction_date": "2021-07-30",
    "amended_from": null,
    "_user_tags": null,
    "_comments": "[{\"comment\": \"<div>Bonjour M XXXXX</div><div><br></div><div><br></div><div>\\u00a0</div><div>XXXXXX est sp\\u00e9ci...\", \"by\": \"user@test.com\", \"name\": \"0185df22a4\"}, {\"comment\": \"<div>M XXXXX </div><div><br></div><div>JDDDDDD..\", \"by\": \"user@test.com\", \"name\": \"d87972479b\"}]",
    "_assign": null,
    "_liked_by": null,
    "_seen": "[\"user@test.com\"]",
    "first_response_time": 120.000000000,
    "expected_closing": null,
    "language": null,
    "opportunity_owner": null,
    "no_of_employees": null,
    "annual_revenue": 0.000000000,
    "industry": null,
    "market_segment": null,
    "website": null,
    "city": null,
    "state": null,
    "country": null,
    "conversion_rate": 0.000000000,
    "base_opportunity_amount": 0.000000000,
    "job_title": null,
    "whatsapp": null,
    "phone": null,
    "phone_ext": null,
    "base_total": 0.000000000,
    "total": 0.000000000
  }
]
```

### CODE REVIEW

**Comment by FHenry** [comment]
patch test will failed until https://github.com/frappe/frappe/pull/17959 will be merge

**Comment by deepeshgarg007** [comment]
@FHenry Can you update the error(Traceback) in the description as well?

**Comment by FHenry** [comment]
> @FHenry Can you update the error(Traceback) in the description as well?

@deepeshgarg007 Hello, done.

**Comment by FHenry** [comment]
@deepeshgarg007 I agree that it should not happen, because if you set an amount on opportunity, the currency come with since last version-12, but this record come from early version-12 upgraded to version-13 a year ago. 
If I got this problem, I will not be the only one, seems to be harmless to make it more secure for all case, even the one that should not happen

File: erpnext/patches/v14_0/update_opportunity_currency_fields.py [hunk 1]
```diff
@@ -16,6 +17,19 @@ def execute():
...
+		if opportunity.currency is None or opportunity.currency == "":
+			opportunity.currency = company_currency
+			frappe.db.set_value(
+				"Opportunity",
+				opportunity.name,
+				{"currency": opportunity.currency},
+				update_modified=False,
+			)
+			click.secho(
+				f' Opportunity `{opportunity.name}` has no currency set. We force it to company currency : `{opportunity.currency}`"\n',
```
Comment (deepeshgarg007) [inline erpnext/patches/v14_0/update_opportunity_currency_fields.py]:
```suggestion
				f' Opportunity `{opportunity.name}` has no currency set. Setting it to company currency as default: `{opportunity.currency}`"\n',
```

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]

**Comment by deepeshgarg007** [comment]
@Mergifyio backport version-14-hotfix

**Pull request merged**